[wip]minio gateway object tagging

BigUstad opened the pull request:
put-object-tagging, get-object-tagging, copy-object, multi-part-upload.

## Description
Gateway S3 object tagging related operations

## Motivation and Context


## How to test this PR?


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

BigUstad committed:
minio gateway object tagging

put-object-tagging, get-object-tagging, copy-object, multi-part-upload.
For s3.

nitisht reviewed (commented):
@BigUstad we should block object operations (Put / Get) with Tagging specific headers in other gateways

BigUstad committed:
minio gateway object tagging
    put-object-tagging, get-object-tagging, copy-object, multi-part-upload.
    For s3.

BigUstad committed:
minio gateway object tagging
s3. put-object-tagging, get-object-tagging, copy-object, multi-part-upload.

kannappanr commented on cmd/api-errors.go:
```diff
@@ -157,6 +157,7 @@ const (
 	ErrObjectLockInvalidHeaders
 	ErrInvalidTagDirective
 	// Add new error codes here.
+	ErrTaggingOperation
```
This code should be added before the comment `//Add new error codes here.`

kannappanr commented on cmd/api-errors.go:
```diff
@@ -851,6 +852,11 @@ var errorCodes = errorCodeMap{
 		HTTPStatusCode: http.StatusBadRequest,
 	},
+	ErrTaggingOperation: {
+		Code:           "InvalidArgument",
```
You can probably use something like `ErrNotImplemented`

BigUstad replied:
ErrTagOperation replaced with ErrNotImplemented

kannappanr commented on cmd/gateway/hdfs/gateway-hdfs.go:
```diff
@@ -735,3 +735,8 @@ func (n *hdfsObjects) AbortMultipartUpload(ctx context.Context, bucket, object,
 	return true
 }
+
+// IsObjectTaggingSupported returns whether tagging is applicable for this layer.
```
same comment everywhere looks redundant

kannappanr commented on cmd/object-handlers.go:
```diff
@@ -808,7 +808,6 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
 		return
 	}
-
```
Please fix the unnecessary new lines removal

BigUstad committed:
minio gateway object tagging

nitisht commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -456,6 +458,13 @@ func (l *s3Objects) PutObject(ctx context.Context, bucket string, object string,
+		if err = l.Client.PutObjectTagging(bucket, object, tagMap); err != nil {
+			return objInfo, minio.ErrorRespToObjectError(err, bucket, object)
+		}
+	}
```
Instead of making another call for `PutObjectTagging`, isn't it better to directly add the tagging headers in `l.Client.PutObject`. This will save one server request.

nitisht commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -489,6 +498,29 @@ func (l *s3Objects) CopyObject(ctx context.Context, srcBucket string, srcObject
+			return objInfo, minio.ErrorRespToObjectError(err, dstBucket, dstObject)
+		}
+	}
+
```
Here also, you can add the tagging related headers in CopyObject API call itself, and let the target S3 server decide on how to handle the tags

nitisht commented on cmd/gateway/s3/gateway-s3-utils.go:
```diff
@@ -45,3 +50,40 @@ func FromGatewayObjectPart(partID int, oi minio.ObjectInfo) (pi minio.PartInfo)
 	}
 }
+
+func makeTagMap(tags tagging.Tagging) map[string]string {
```
This method can be added to the `tagging` package (`pkg/bucket/object/tagging`)

kannappanr commented on cmd/object-handlers.go:
```diff
@@ -1012,8 +1012,11 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 	}
-
-	if tags != "" {
+	if !IsStringEqual(tags, "") && !objectAPI.IsObjectTaggingSupported() {
```
The `IsObjectTaggingSupported` can be brought inside the check for tag being empty

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -452,7 +453,14 @@ func (l *s3Objects) GetObjectInfo(ctx context.Context, bucket string, object str
 
-	oi, err := l.Client.PutObject(bucket, object, data, data.Size(), data.MD5Base64String(), data.SHA256HexString(), minio.ToMinioClientMetadata(opts.UserDefined), opts.ServerSideEncryption)
+	var tagMap map[string]string
+	if !minio.IsStringEqual(opts.UserDefined["X-Amz-Tagging"], "") {
```
You can check for the existence of this key, instead of this check

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -452,7 +453,14 @@ func (l *s3Objects) GetObjectInfo(ctx context.Context, bucket string, object str
-	oi, err := l.Client.PutObject(bucket, object, data, data.Size(), data.MD5Base64String(), data.SHA256HexString(), minio.ToMinioClientMetadata(opts.UserDefined), opts.ServerSideEncryption)
+	var tagMap map[string]string
+	if !minio.IsStringEqual(opts.UserDefined["X-Amz-Tagging"], "") {
+		opts.UserTags = opts.UserDefined["X-Amz-Tagging"]
```
If `USerTags` will not be used, why set it?

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -474,6 +482,14 @@ func (l *s3Objects) CopyObject(ctx context.Context, srcBucket string, srcObject
 	srcInfo.UserDefined["x-amz-metadata-directive"] = "REPLACE"
 	srcInfo.UserDefined["x-amz-copy-source-if-match"] = srcInfo.ETag
+	// See comment above. Tags will be applied on destination object.
+	if !minio.IsStringEqual(srcInfo.UserDefined["X-Amz-Tagging"], "") {
```
instead of checking for empty string, you can check for the presence of this key

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -474,6 +482,14 @@ func (l *s3Objects) CopyObject(ctx context.Context, srcBucket string, srcObject
+		if minio.IsStringEqual(srcInfo.UserDefined["X-Amz-Tagging-Directive"], "") {
+			srcInfo.UserDefined["x-amz-tagging-directive"] = "REPLACE"
+		}
+		srcInfo.UserTags = srcInfo.UserDefined["X-Amz-Tagging"]
```
Is the tags getting used anywhere?

kannappanr replied:
without `REPLACE` does it make sense to have the `tags` sent over from the client?

BigUstad replied:
I've removed setting of UserTags. Testing showed it to work with both S3 & minio server. REPLACE directive is needed by the functionality.

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -635,6 +656,50 @@ func (l *s3Objects) DeleteBucketPolicy(ctx context.Context, bucket string) error
+}
+
+func (l *s3Objects) DeleteObjectTag(ctx context.Context, bucket, object string) error {
+	var err error
```
```suggestion
```

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -635,6 +656,50 @@ func (l *s3Objects) DeleteBucketPolicy(ctx context.Context, bucket string) error
+
+func (l *s3Objects) DeleteObjectTag(ctx context.Context, bucket, object string) error {
+	var err error
+	if err = l.Client.RemoveObjectTagging(bucket, object); err != nil {
```
```suggestion
	if err := l.Client.RemoveObjectTagging(bucket, object); err != nil {
```

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -523,11 +539,16 @@ func (l *s3Objects) ListMultipartUploads(ctx context.Context, bucket string, pre
-	opts := miniogo.PutObjectOptions{UserMetadata: o.UserDefined, ServerSideEncryption: o.ServerSideEncryption}
+	var tags map[string]string
+	if tags, err = getTagMap(o.UserDefined["X-Amz-Tagging"]); err != nil {
+		return uploadID, minio.ErrorRespToObjectError(err, bucket, object)
```
What sort of objecterror does this return? Have you verified that this matches with what AWS S3 returns?

BigUstad replied:
The errors are the ones that might be returned by tagging.Validate or XML decoder functions in tagging pkg. tagging.Validate() returns errors like ErrTooManyTags, ErrInvalidTag .etc. It is compatible with S3, I'd presume.

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -635,6 +656,50 @@ func (l *s3Objects) DeleteBucketPolicy(ctx context.Context, bucket string) error
+		return tagging.Tagging{}, minio.ErrorRespToObjectError(err, bucket, object)
+	}
+	if tags, err = tagging.ParseTagging(strings.NewReader(tagStr)); err != nil {
+		return tagging.Tagging{}, minio.ErrorRespToObjectError(err, bucket, object)
```
what are the error scenarios here? Have you verified if gateway returns same errors in that scenario as AWS S3?

kannappanr commented on cmd/gateway/s3/gateway-s3.go:
```diff
@@ -474,6 +482,14 @@ func (l *s3Objects) CopyObject(ctx context.Context, srcBucket string, srcObject
 	srcInfo.UserDefined["x-amz-copy-source-if-match"] = srcInfo.ETag
+	// See comment above. Tags will be applied on destination object.
+	if !minio.IsStringEqual(srcInfo.UserDefined["X-Amz-Tagging"], "") {
+		if minio.IsStringEqual(srcInfo.UserDefined["X-Amz-Tagging-Directive"], "") {
```
check for the presence of this key

BigUstad committed:
[wip]minio gateway object tagging
